File: tests/cpp/TestNeuralNetwork.cpp
#include <gtest/gtest.h>

#include <optlib/core/Differentiation.hpp>
#include <optlib/core/nn/NeuralNetwork.hpp>

#include <vector>

namespace {

optlib::Matrix XorFeatures()
{
    const std::vector<double> values{0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 1.0};
    return optlib::Matrix(4, 2, values);
}

optlib::Vector XorTargets()
{
    return optlib::Vector{0.0, 1.0, 1.0, 0.0};
}

} // namespace

TEST(NeuralNetworkTest, BackpropMatchesFiniteDifference)
{
    const auto features = XorFeatures();
    const auto targets = XorTargets();
    optlib::BinaryMlpConfig config;
    config.InputDim = 2;
    config.HiddenDim = 3;
    config.Activation = optlib::ActivationType::Tanh;
    auto parameters =
        optlib::InitializeBinaryMlpParameters(config, optlib::InitializationType::Xavier, 7);
    optlib::Vector gradient(optlib::BinaryMlpParameterCount(config));
    const auto loss = optlib::BinaryMlpLossAndGradient(
        parameters.Span(), features, targets.Span(), config, gradient.Span());
    EXPECT_GT(loss, 0.0);

    auto numericGradient = optlib::ComputeGradient(
        [&features, &targets, &config](std::span<const double> values) {
            optlib::Vector scratch(optlib::BinaryMlpParameterCount(config));
            return optlib::BinaryMlpLossAndGradient(
                values, features, targets.Span(), config, scratch.Span());
        },
        parameters.Span(),
        optlib::DifferentiationOptions{optlib::DifferentiationScheme::Central, 1e-5});

    for (std::size_t index = 0; index < gradient.Size(); ++index) {
        EXPECT_NEAR(gradient[index], numericGradient[index], 1e-5);
    }
}

TEST(NeuralNetworkTest, TrainsXorWithAdam)
{
    const auto features = XorFeatures();
    const auto targets = XorTargets();
    optlib::BinaryMlpTrainConfig config;
    config.Network.InputDim = 2;
    config.Network.HiddenDim = 8;
    config.Network.Activation = optlib::ActivationType::Tanh;
    config.Method = optlib::FirstOrderMethod::Adam;
    config.Optimizer.LearningRate = 0.05;
    config.Optimizer.Criteria.MaxIterations = 5000;
    config.Optimizer.Criteria.GradientTolerance = 1e-5;
    config.Optimizer.Criteria.StepTolerance = 0.0;
    config.Optimizer.Criteria.FunctionTolerance = 0.0;
    config.Optimizer.StoreTrajectory = false;
    config.Seed = 11;

    const auto result = optlib::TrainBinaryMlp(features, targets.Span(), config);
    EXPECT_LT(result.Loss, 0.05);
    EXPECT_GT(result.F1, 0.99);
}

File: tests/python/test_neural_network.py
from __future__ import annotations

import numpy as np

import optlib


def test_binary_mlp_loss_gradient_shapes() -> None:
    features, targets = optlib.make_xor()
    parameters = optlib.InitializeBinaryMlpParameters(2, 4, seed=7)
    result = optlib.BinaryMlpLossAndGradient(parameters, features, targets, 2, 4)

    assert result["loss"] > 0.0
    assert result["gradient"].shape == parameters.shape
    assert optlib.BinaryMlpParameterCount(2, 4) == parameters.shape[0]


def test_mlp_classifier_fits_xor() -> None:
    features, targets = optlib.make_xor()
    classifier = optlib.MLPClassifier(
        hidden_dim=8,
        learning_rate=0.05,
        max_iter=5000,
        gradient_tolerance=1e-5,
        seed=11,
    ).fit(features, targets)

    assert classifier.loss_ is not None and classifier.loss_ < 0.05
    assert classifier.score(features, targets) > 0.99
    np.testing.assert_array_equal(classifier.predict(features), targets.astype(np.int64))


def test_mlp_classifier_two_moons_sanity() -> None:
    features, targets = optlib.make_two_moons(samples=120, noise=0.05, seed=4)
    classifier = optlib.MLPClassifier(
        hidden_dim=12,
        learning_rate=0.03,
        max_iter=4000,
        gradient_tolerance=1e-5,
        seed=5,
    ).fit(features, targets)

    assert classifier.score(features, targets) > 0.9
